### PR TITLE
Added the aws resource glue resource policy

### PR DESCRIPTION
**Pull request by stijndehaes**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
new resource: aws_glue_resource_policy
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueResourcePolicy_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Basic -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Basic
2019/10/03 07:49:43 [DEBUG] Unable to read account ID from test provider: unconfigured provider
=== PAUSE TestAccAWSGlueResourcePolicy_Basic
=== CONT  TestAccAWSGlueResourcePolicy_Basic
--- PASS: TestAccAWSGlueResourcePolicy_Basic (27.04s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       27.095s



$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueResourcePolicy_Update'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Update -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Update
2019/10/03 07:42:22 [DEBUG] Unable to read account ID from test provider: unconfigured provider
2019/10/03 07:42:22 [DEBUG] Unable to read account ID from test provider: unconfigured provider
=== PAUSE TestAccAWSGlueResourcePolicy_Update
=== CONT  TestAccAWSGlueResourcePolicy_Update
--- PASS: TestAccAWSGlueResourcePolicy_Update (48.10s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       48.154s

```

If someone could help me with the test setup of the acceptance tests that would be nice. I added the environment variable `AWS_PROFILE` and the acceptance tests get access to aws. However it can not read the account id from the test provider. So I manually added in the policy to run the test and removed it after account. Anyone can help me with my setup?

Also since this is my first PR please help me with anything I forgot to add.

### CODE REVIEW

**Comment by zbstof** [comment]
@stijndehaes, is failing build a problem?

**Comment by lucienfregosi** [comment]
Hi @stijndehaes 
Do you plan to fix Travis Integration and merge this PR ? Actually we would love to use this feature.
Maybe we can help you ? 
Many Thanks

**Comment by stijndehaes** [comment]
@lucienfregosi Thank you for reminding me :) Rebased and will follow up the PR tomorrow and see if I can fix the Travis issues. I can't get this PR merged though we would have to get the attention form someone from Hashicorp for that. Feel free to review my PR in the mean time :)

**Comment by stijndehaes** [comment]
Build got fixed ready for review :)

File: website/docs/r/glue_resource_policy.html.markdown [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+### Non-VPC Connection
+
+```hcl
+data "aws_iam_policy_document" "glue-example-policy" {
+  statement {
+    actions = [
+      "glue:CreateTable",
+    ]
+    resources = ["arn:aws:glue:AWS_REGION:ACCOUNT_ID:*"]
```
Comment (bflad) [inline website/docs/r/glue_resource_policy.html.markdown]: Recommend updating this example to match the updated test configurations with the data sources. 👍

File: website/docs/r/glue_resource_policy.html.markdown [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  Provides a resource to configure the aws glue resource policy.
+---
+
+# Resource: aws_glue_resource_policy
+
+Provides a Glue resource policy. Only one can exist per region.
+
+## Example Usage
+
+### Non-VPC Connection
```
Comment (bflad) [inline website/docs/r/glue_resource_policy.html.markdown]:
```suggestion
```

**Comment by stijndehaes** [comment]
@bflad I updated the code an reran the acceptance tests:

```

TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Basic -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Basic
=== PAUSE TestAccAWSGlueResourcePolicy_Basic
=== CONT  TestAccAWSGlueResourcePolicy_Basic
--- PASS: TestAccAWSGlueResourcePolicy_Basic (39.43s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       40.713s
```

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueResourcePolicy_Update'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Update -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Update
=== PAUSE TestAccAWSGlueResourcePolicy_Update
=== CONT  TestAccAWSGlueResourcePolicy_Update
--- PASS: TestAccAWSGlueResourcePolicy_Update (60.31s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       61.593s
```
Everything seems to be working now :) Thank you very much for the pointers on the tests and thanks for the review.

File: aws/resource_aws_glue_resource_policy_test.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        "glue:CreateTable"
+      ],
+      "Principal" : {
+         "AWS": "*"
+       },
+      "Resource" : "arn:aws:glue:%s:%s:*"
+    }
+  ]
+}`, testAccGetRegion(), testAccGetAccountID())
+}
```
[the remark this answers is not in the document]
Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy_test.go]: Thank you for the pointers they helped a ton!

**Comment by stijndehaes** [comment]
Rebased again

**Comment by stijndehaes** [comment]
Rebased this branch to the latest master commit

**Comment by stijndehaes** [comment]
Also reran acceptance tests:

```
➜  terraform-provider-aws git:(feature/aws-glue-data-catalog-resource-policy) ✗ TF_ACC=1 make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueResourcePolicy_Basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Basic -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Basic
=== PAUSE TestAccAWSGlueResourcePolicy_Basic
=== CONT  TestAccAWSGlueResourcePolicy_Basic
--- PASS: TestAccAWSGlueResourcePolicy_Basic (27.86s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       32.897s
➜  terraform-provider-aws git:(feature/aws-glue-data-catalog-resource-policy) ✗ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueResourcePolicy_Update'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlueResourcePolicy_Update -timeout 120m
=== RUN   TestAccAWSGlueResourcePolicy_Update
=== PAUSE TestAccAWSGlueResourcePolicy_Update
=== CONT  TestAccAWSGlueResourcePolicy_Update
--- PASS: TestAccAWSGlueResourcePolicy_Update (45.73s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       49.653s

```

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by DrFaust92** [comment]
Hey @stijndehaes, do you mind rebasing again and fixing things up according to the bot message above?

on another note, can you add the `ResourceArn` to set the policy for a specific resource? optionally support for `EnableHybrid` can also be added

**Comment by stijndehaes** [comment]
> ResourceArn

Hey I will update my PR and I'll have to look into `ResourceArn` didn't know that that existed.

**Comment by DrFaust92** [comment]
Hey @stijndehaes, are you still lokking into this?

**Comment by stijndehaes** [comment]
> Hey @stijndehaes, are you still lokking into this?

I will make the time to do this next week :) It's not hard to add I just have to make sure I make the right integration test

**Comment by stijndehaes** [comment]
@DrFaust92 I had a look at adding `ResourceArn` support tonight however it looks like this is actually unsupported for glue. For example this aws cli command (pleas don't forget to replace ACCOUNTID):

```
aws glue put-resource-policy --region us-west-2 --resource-arn arn:aws:glue:us-west-2:ACCOUNTID:catalog --policy-in-json '{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Principal": {
        "AWS": ["*"]
      },
      "Effect": "Allow",
      "Action": [
        "glue:CreateTable"
      ],
      "Resource": [
        "arn:aws:glue:us-west-2:ACCOUNTID:catalog"
      ]
    }
  ]
}'
```

results in the following error:
```An error occurred (InvalidInputException) when calling the PutResourcePolicy operation: Glue resource level policy is not currently supported```

So I don't think this field is supported yet for glue. I also don't find any mention anywere for setting resource policies on database or tables in the documentation.

**Comment by DrFaust92** [comment]
Hey @stijndehaes, if so whats the use case for the general policy?

**Comment by ferrouswheel** [comment]
@DrFaust92 I've been watching this PR for a wee while as I need to allow glue catalog access from another AWS account.

**Comment by Thiago-Dantas** [comment]
@DrFaust92  same as @ferrouswheel here

File: aws/resource_aws_glue_resource_policy_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	return fmt.Sprintf(`
+data "aws_caller_identity" "current" {}
+
+data "aws_partition" "current" {}
+
+data "aws_region" "current" {}
+
+data "aws_iam_policy_document" "glue-example-policy" {
+  statement {
+    actions = ["%s"]
```
Comment (DrFaust92) [inline aws/resource_aws_glue_resource_policy_test.go:62]: I believe that the lint issue is to add more spaces before `=` to align with `resources` below it and it looks good to go.

**Comment by stijndehaes** [comment]
> Hey @stijndehaes, if so whats the use case for the general policy?

The general policy just allows you to set access to the catalog and all the databases/tables that are under the catalog. 
So it allows you to say another account can read a table of this catalog. Or all tables of this catalog.
For more information you can have a look here: https://docs.aws.amazon.com/glue/latest/dg/glue-resource-policies.html

File: aws/resource_aws_glue_resource_policy_test.go (hunk 4 above)
Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy_test.go]: Good catch I really had no clue looking at the logs

Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy_test.go]: Fixed it now I get the same exact error as on master 😅 
I'll check again tomorrow if it's fixed in master and rebase if it's fixed

Comment (DrFaust92) [reply aws/resource_aws_glue_resource_policy_test.go]: oh, good. its ok. ill run test now to verify. it will be fixed on master.

**Comment by DrFaust92** [comment]
Hey @stijndehaes, when you have time I just noticed a `disappears` test is missing. see https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md#disappears-acceptance-tests

File: aws/resource_aws_glue_resource_policy.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateFunc:     validation.StringIsJSON,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+			},
+		},
+	}
+}
+
+func resourceAwsGlueResourcePolicyCreate(d *schema.ResourceData, meta interface{}) error {
```
Comment (DrFaust92) [inline aws/resource_aws_glue_resource_policy.go:33]: at a closer look, I think we can compress the create/update functions to a single `resourceAwsGlueResourcePolicyPut` see `resourceAwsApiGatewayGatewayResponse` resource for an example

File: aws/resource_aws_glue_resource_policy.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	conn := meta.(*AWSClient).glueconn
+
+	_, err := conn.PutResourcePolicy(&glue.PutResourcePolicyInput{
+		PolicyInJson:          aws.String(d.Get("policy").(string)),
+		PolicyExistsCondition: aws.String(glue.ExistConditionNotExist),
+	})
+	if err != nil {
+		return fmt.Errorf("error putting policy request: %s", err)
+	}
+	d.SetId(meta.(*AWSClient).accountid)
```
Comment (DrFaust92) [inline aws/resource_aws_glue_resource_policy.go:43]: isn't glue regional? it may be better to use region here.

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
some more comments,

After running tests I realised we need to serialize tests: see [example](https://github.com/terraform-providers/terraform-provider-aws/blob/167989510020c93cf89fac5c2c4ddcd593c9dfa6/aws/resource_aws_securityhub_test.go)

File: aws/resource_aws_glue_resource_policy.go (hunk 5 above)
Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy.go:33]: There is a small difference the `PolicyExistsCondition` condition changes but I wrapped it in a function that now looks something like this: 
`Create: resourceAwsGlueResourcePolicyPut(glue.ExistConditionNotExist),`

**Comment by stijndehaes** [comment]
> Hey @stijndehaes, when you have time I just noticed a `disappears` test is missing. see https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md#disappears-acceptance-tests



> some more comments,
> 
> After running tests I realised we need to serialize tests: see [example](https://github.com/terraform-providers/terraform-provider-aws/blob/167989510020c93cf89fac5c2c4ddcd593c9dfa6/aws/resource_aws_securityhub_test.go)

Created the serialize tests, also changed the tests to `resource.Test` instead of paralleltest since they are editing the same resource.

Output of integration tests:

```bash
➜  terraform-provider-aws git:(feature/aws-glue-data-catalog-resource-policy) ✗ make testacc TESTARGS='-run=TestAccAWSGlue_serial'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlue_serial -timeout 120m
=== RUN   TestAccAWSGlue_serial
=== RUN   TestAccAWSGlue_serial/ResourcePolicy
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/basic
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/update
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/disappears
--- PASS: TestAccAWSGlue_serial (103.45s)
    --- PASS: TestAccAWSGlue_serial/ResourcePolicy (103.45s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/basic (29.37s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/update (48.87s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/disappears (25.21s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       105.155s
```

**Comment by stijndehaes** [comment]
Fixed the last linting error everything is green now :)

**Review by DrFaust92 — APPROVED** [review]
LGTM

```
--- PASS: TestAccAWSGlue_serial (141.95s)    --- PASS: TestAccAWSGlue_serial/ResourcePolicy (141.95s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/basic (39.27s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/update (69.54s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/disappears (33.13s)
```

File: aws/resource_aws_glue_resource_policy_test.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      "Action" : [
+        "%s"
+      ],
+      "Principal" : {
+         "AWS": "*"
+       },
+      "Resource" : "arn:aws:glue:%s:%s:*"
+    }
+  ]
+}`, action, testAccGetRegion(), testAccGetAccountID())
```
Comment (YakDriver) [inline aws/resource_aws_glue_resource_policy_test.go:27]:
```suggestion
      "Resource" : "arn:%s:glue:%s:%s:*"
    }
  ]
}`, action, testAccGetPartition(), testAccGetRegion(), testAccGetAccountID())
```

**Review by YakDriver — CHANGES_REQUESTED** [review]

File: aws/resource_aws_glue_resource_policy_test.go (hunk 7 above)
Comment (breathingdust) [reply aws/resource_aws_glue_resource_policy_test.go:27]: This change is to prevent test failures when run in GovCloud, otherwise this is good to go.

Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy_test.go]: Done

Comment (stijndehaes) [reply aws/resource_aws_glue_resource_policy_test.go]: Reran tests after the change:
```
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlue_serial/ResourcePolicy -timeout 120m
=== RUN   TestAccAWSGlue_serial
=== RUN   TestAccAWSGlue_serial/ResourcePolicy
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/basic
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/update
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/disappears
--- PASS: TestAccAWSGlue_serial (106.69s)
    --- PASS: TestAccAWSGlue_serial/ResourcePolicy (106.69s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/basic (31.54s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/update (49.40s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/disappears (25.74s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       108.422s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @stijndehaes!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlue_serial'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlue_serial -timeout 120m
=== RUN   TestAccAWSGlue_serial
=== RUN   TestAccAWSGlue_serial/ResourcePolicy
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/disappears
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/basic
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/update
--- PASS: TestAccAWSGlue_serial (59.41s)
    --- PASS: TestAccAWSGlue_serial/ResourcePolicy (59.41s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/disappears (14.40s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/basic (15.77s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/update (29.25s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	60.858s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlue_serial'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGlue_serial -timeout 120m
=== RUN   TestAccAWSGlue_serial
=== RUN   TestAccAWSGlue_serial/ResourcePolicy
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/disappears
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/basic
=== RUN   TestAccAWSGlue_serial/ResourcePolicy/update
--- PASS: TestAccAWSGlue_serial (37.45s)
    --- PASS: TestAccAWSGlue_serial/ResourcePolicy (37.45s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/disappears (9.15s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/basic (10.65s)
        --- PASS: TestAccAWSGlue_serial/ResourcePolicy/update (17.65s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	38.817s
```

**Pull request merged**